### DISCUSSION TITLE
Question: Nullable reference-friendly libraries

### DISCUSSION THREAD

**Discussion by mythgarr** [dotnet/csharplang#1851] [Q&A]
Libraries written in C# have an obvious path for migrating to nullable reference syntax (once C# 8.0 support is available). My question is how libraries authored in other languages (C++/CLI in particular) are expected to be annotated to inform the C# compiler that
1) The library has opted into explicitly annotating all nullable references
2) That a given return value or argument is nullable

The Roslyn preview creates an internal NullableAttribute class within the assembly to indicate the first. Is this planned to be promoted to a core attribute prior to release of this language feature? Is attribute annotation likely to be the final solution?

**Comment by YairHalberstadt** [comment]
You can define an internal NullableAttribute manually yourself and it will work. However the spec of how to apply it for generic types can be complicated to say the least.

**Reply by jnm2** [reply]
https://github.com/dotnet/roslyn/blob/master/docs/features/nullable-metadata.md
